### ISSUE TITLE
Custom scalar isn't being respected

### ISSUE THREAD

**Issue by solidcell** [apollographql/apollo-ios#3371]
### Summary

https://www.loom.com/share/c6faff0cb93f4c9b85342cdb9053e52c?sid=ed020aee-67eb-4831-9832-4b5645a4535a

I've defined a custom scalar `BasicScalar` which is seemingly not being respected/used for a field which is defined as `result: BasicScalar!`. Instead, the `DataDict` ends up having `result: AnyHashable(Bool)`. So instead of it decoding the query so that `result` is `BasicScalar` (`.bool(true)`), it's decoding `result` as plain `true` and not using the `BasicScalar` type at all, as the GraphQL schema specifies. What am I doing wrong? My schema, operation, and custom scalar all seem to be defined accordingly, yet..

### Version

1.9.0

### Steps to reproduce the behavior

See the summary.

**Comment by calvincestari** [comment]
Hi @solidcell - is the data value being sent as regular JSON bool? Something like this:
```
{
  result: true
}
```

**Comment by solidcell** [comment]
Yea. In this case it's `true` but it could be for example `54`. I've defined `init(_jsonValue value: JSONValue)` so I'm not sure why it's being decoded as a plain `Bool` for this response.

**Comment by calvincestari** [comment]
Are you able to share the definition for `BasicScalar`? Does the behaviour change if the value is `54` for example, i.e.: something other than a boolean value?

**Comment by solidcell** [comment]
https://www.loom.com/share/b8b4ae318a94464a96b63f1ddb47bc48?sid=e9c3f788-be91-469d-8e41-be67a06850a9

**Comment by calvincestari** [comment]
OK, I think the issue is your implementation of `_asAnyHashable`. It in turn calls the private function `hashable` which is returning the base type such as `Bool` or `Int`. I believe it's that value that ends up getting stored in `DataDict`. You shouldn't need to define `_asAnyHashable` at all. Essentially you're turning your custom scalar back into the base type.

[Here](https://github.com/apollographql/apollo-ios/files/15100165/SampleProject-Issue3371.zip) is a sample project which defines a custom scalar pretty similar to yours just with less variants to it. The test passes with the custom scalar type being returned as expected. In the custom scalar file there is a commented out portion with the same `_asAnyHashable` and `hashable` functions that you have, and if you uncomment that section you'll get failures in the test with the same error you're seeing.

tl;dr - remove your `_asAnyHashable` function and it should work as expected.

**Comment by solidcell** [comment]
That was it, thanks! I was waiting to trim code until things started working 🤦 

Appreciate your help!
